rust/gelato/src/chains.rs

```diff
@@ -0,0 +1,148 @@
+// in string format is useful for CLI usage so that we don't
+// have to remember chain IDs and can instead refer to names.
+
+use ethers::types::{Address, U256};
+use std::{fmt, str::FromStr};
+
+use crate::err::GelatoError;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Chain {
+    Mainnet,
+    Ropsten,
+    Rinkeby,
+    Goerli,
+    Kovan,
+    XDai,
+    Polygon,
+    PolygonMumbai,
+    Avalanche,
+    AvalancheFuji,
+    Sepolia,
+    Moonbeam,
+    MoonbeamDev,
+    Moonriver,
+}
+
+impl fmt::Display for Chain {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "{:?}", self)
+    }
+}
+
+impl FromStr for Chain {
+    type Err = GelatoError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "mainnet" => Ok(Chain::Mainnet),
+            "ropsten" => Ok(Chain::Ropsten),
+            "rinkeby" => Ok(Chain::Rinkeby),
+            "goerli" => Ok(Chain::Goerli),
+            "kovan" => Ok(Chain::Kovan),
+            "xdai" => Ok(Chain::XDai),
+            "polygon" => Ok(Chain::Polygon),
+            "polygonmumbai" => Ok(Chain::PolygonMumbai),
+            "avalanche" => Ok(Chain::Avalanche),
+            "avalanchefuji" => Ok(Chain::AvalancheFuji),
+            "sepolia" => Ok(Chain::Sepolia),
+            "moonbeam" => Ok(Chain::Moonbeam),
+            "moonbeamd" => Ok(Chain::MoonbeamDev),
+            "moonriver" => Ok(Chain::Moonriver),
+            _ => Err(GelatoError::UnknownChainNameError(String::from(s))),
+        }
+    }
+}
+
+impl From<Chain> for u32 {
+    fn from(chain: Chain) -> Self {
+        match chain {
+            Chain::Mainnet => 1,
+            Chain::Ropsten => 3,
+            Chain::Rinkeby => 4,
+            Chain::Goerli => 5,
+            Chain::Kovan => 42,
+            Chain::XDai => 100,
+            Chain::Polygon => 137,
+            Chain::PolygonMumbai => 80001,
+            Chain::Avalanche => 43114,
+            Chain::AvalancheFuji => 43113,
+            Chain::Sepolia => 11155111,
+            Chain::Moonbeam => 1287,
+            Chain::MoonbeamDev => 1281,
+            Chain::Moonriver => 1285,
+        }
+    }
+}
+
+impl From<Chain> for U256 {
+    fn from(chain: Chain) -> Self {
+        u32::from(chain).into()
+    }
+}
+
+impl From<Chain> for u64 {
+    fn from(chain: Chain) -> Self {
+        u32::from(chain).into()
+    }
+}
+
+impl Chain {
+    // We also have to provide hardcoded verification contract addresses
+    // for Gelato-suppored chains, until a better / dynamic approach
+    // becomes available.
+    //
+    // See `getRelayForwarderAddrss()` in the SDK file
+    // `@gelatonetwork/gelato-relay-sdk/package/dist/constants/index.js`.
+    pub fn relay_fwd_addr(&self) -> Result<Address, GelatoError> {
+        match self {
+            Chain::Rinkeby => Ok(Address::from_str(
+                "9B79b798563e538cc326D03696B3Be38b971D282",
+            )?),
```

Comment:
Is our bar for 'ok to just `unwrap`' just 'code mistake'? I get what you're saying, but I wonder if we ought to consider a tighter standard in certain contexts where panics are costly and deterministic and would clearly affect liveness of the network. Here is an example of the category of scenario I'd worry about: an agent is getting lots of traffic to forward messages from chain A to chain B. It has been running for a long time without issue. Suddenly somebody on chain A posts a message to be forwarded to chain C. For the first time, the agent consults rust/gelato/src/chains.rs Chain::from_str() which consults the hardcoded string address. For whatever reason it is (let's say) an odd number of characters due to a copy/paste error, and we didn't/don't have universal line coverage for unit tests, and so it won't parse and we panic. When the agent comes back up, it will find the same work to deliver the message from A to C and panic again, and will do so idnefinitely until we are alerted. That is not ideal in its own right, but more sadly, it would halt transmission of messages from A->B which had been heretofore working smoothly. Wouldn't we prefer to return an error for which we had monitoring that would alert us / the right person as to the condition and have the agent continue processing what it can meanwhile?

Reply:
Ideally we are testing our code and will run into them there. I see what you are saying but my personal heuristic for panics is:

1) Anything we know for a fact cannot happen (usually best to document why it is impossible if it is not overly obvious to a casual observer)
2) code error that really should just be fixed if it ever occurs for any reason
3) anything that we don't think will actually happen **and** we are in a hurry (this should be an `except` not an `unwrap` so we can pin it down and fix if it turns out to actually happen)

Reply:
Those are fine heuristics, but do you have any thoughts about the liveness concern I raised?

Reply:
The agents crash and restart as it stands, it's not great but that is why I suggest focusing on code errors because then it's like an instant bug report. In this exact case if you think keeping it as you have it is best, I am not giving a hard and fast "make this an unwrap".

Reply:
Alright, I think I'm going to leave as-is for now if that's cool (taking your 'not giving a hard and fast' as a signal that it is). One additional reason I'm ok with returning an error (aside from liveness concerns) is that since we handle the default case for an UnknownRelayForwardAddress, any caller has to handle the possibility of an error anyway, so I'm not sure we'd be getting any huge simplicity gain.